### PR TITLE
Contribution guide and PR template

### PR DESCRIPTION
**Pull request by jelysn**

### CODE REVIEW

File: CODING_GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+## API & Design
+
+- Code must be well structured:
+  - packages must have a limited responsibility (different concerns can go to different packages),
+  - types must be easy to compose,
+  - think about maintainbility and testability.
+- "Depend upon abstractions, [not] concretions".
+- Try to limit the number of methods you are exposing. It's easier to expose something later than to hide it.
+- Take advantage of `internal` package concept.
+- Follow agreed-upon design patterns and naming conventions.
```
Comment (cosmic-vagabond) [inline CODING_GUIDELINES.md:14]: @jelysn I think we need to clarify that maybe link to the golang coding best practices or borrow something from the cosmos dev community.

File: CODING_GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+- Avoid unnecessary operations or memory allocations.
+
+Security:
+
+- Pay proper attention to exploits involving:
+  - gas usage
+  - transaction verification and signatures
+  - malleability
+  - code must be always deterministic
```
Comment (cosmic-vagabond) [inline CODING_GUIDELINES.md:31]: maybe we can provide some examples of codes that could affect the order in which the flow is executed such as non deterministic list orders, random number generators etc.

File: CODING_GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+- Thread safety. If some functionality is not thread-safe, or uses something that is not thread-safe, then clearly indicate the risk on each level.
```
Comment (cosmic-vagabond) [inline CODING_GUIDELINES.md:32]: here as well I would add some examples of scenarios where a code could be not thread-safe.

File: CODING_GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+## Automated Tests
+
+Make sure your code is well tested:
+
+- Provide unit tests for every unit of your code if possible. Unit tests are expected to comprise 70%-80% of your tests.
```
Comment (cosmic-vagabond) [inline CODING_GUIDELINES.md:38]: @jelysn on top of the guideline we should enforce that with tools, a PR cannot be merged unless it has at least 70-80% test coverage for example.

**Comment by cosmic-vagabond** [comment]
@jelysn what do you think of adding those additional sections in the PR template such as:

```
### Screenshots and Videos
Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.
```

and:

```
### What has Changed?
What specific problem were you aiming to address, and how did you successfully resolve it? If tests were not uploaded for this pull request or if coverage decreased, please provide an explanation for the change.
```

and:

```
### Deployment Notes
Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.
```

File: CODING_GUIDELINES.md (hunk 4 above)
Comment (jelysn) [reply CODING_GUIDELINES.md:38]: I think we will need to integrate code test coverage bot on CI.

**Comment by jelysn** [comment]
> ### Screenshots and Videos
> Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process

Added these items based on your comments on PR templates

File: CODING_GUIDELINES.md (hunk 3 above)
Comment (jelysn) [reply CODING_GUIDELINES.md]: Added example

File: CODING_GUIDELINES.md (hunk 2 above)
Comment (jelysn) [reply CODING_GUIDELINES.md]: Added example

File: CODING_GUIDELINES.md (hunk 1 above)
Comment (jelysn) [reply CODING_GUIDELINES.md]: Added reference link

File: CODING_GUIDELINES.md (hunk 4 above)
Comment (cosmic-vagabond) [reply CODING_GUIDELINES.md:41]: Agreed added to issues#115

**Review by cosmic-vagabond — APPROVED** [review]
LGTM thanks a lot

**Pull request merged**